usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -1101,19 +1139,20 @@ metaslab_block_find(avl_tree_t *t, uint64_t start, uint64_t size)
  * tree looking for a block that matches the specified criteria.
  */
 static uint64_t
-metaslab_block_picker(avl_tree_t *t, uint64_t *cursor, uint64_t size,
-    uint64_t align)
+metaslab_block_picker(metaslab_t *msp, avl_tree_t *t, uint64_t *cursor,
+    uint64_t size, uint64_t align)
 {
 	range_seg_t *rs = metaslab_block_find(t, *cursor, size);
 
-	while (rs != NULL) {
+	for (; rs != NULL; rs = AVL_NEXT(t, rs)) {
 		uint64_t offset = P2ROUNDUP(rs->rs_start, align);
 
-		if (offset + size <= rs->rs_end) {
+		if (offset + size <= rs->rs_end &&
+		    !metaslab_check_trim_conflict(msp, &offset, size, align,
+		    rs->rs_end)) {
```

Comment:
1. That makes sense.  Seems like you'd need to limit the number of metaslabs that are being concurrently trimmed, on each vdev.

2. Yeah.  Rather than an on-disk format change, I'd suggest changing `metaslab_condense()` to know that both `ms_tree` and `ms_trimming_ts` hold available space.

3. Fair enough.

For metaslab_ff_alloc, I mean remove it and anything that then becomes dead code (I think just `metaslab_ff_ops`, and the `align` argument to `metaslab_block_picker`).